indexer/pkg/indexer/indexer.go

```diff
@@ -0,0 +1,578 @@
+	"github.com/elastic/go-elasticsearch/v8/esutil"
+	"github.com/elastic/go-elasticsearch/v8/typedapi/core/search"
+	estypes "github.com/elastic/go-elasticsearch/v8/typedapi/types"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/lmittmann/w3"
+	"github.com/lmittmann/w3/module/eth"
+)
+
+type BlockchainIndexer struct {
+	ethClient                EthereumClient
+	esClient                 ElasticsearchClient
+	blockChan                chan *types.Block
+	txChan                   chan *types.Transaction
+	indexInterval            time.Duration
+	lastForwardIndexedBlock  *big.Int
+	lastBackwardIndexedBlock *big.Int
+	logger                   *slog.Logger
+}
+
+type EthereumClient interface {
+	GetBlocks(ctx context.Context, blockNums []*big.Int) ([]*types.Block, error)
+	BlockByNumber(ctx context.Context, number *big.Int) (*types.Block, error)
+	BlockNumber(ctx context.Context) (*big.Int, error)
+	TxReceipts(ctx context.Context, txHashes []string) (map[string]*types.Receipt, error)
+}
+
+type ElasticsearchClient interface {
+	Index(ctx context.Context, index string, document interface{}) error
+	Search(ctx context.Context, index string, query *estypes.Query) (*search.Response, error)
+	GetLastIndexedBlock(ctx context.Context, direction string) (*big.Int, error)
+	CreateIndices(ctx context.Context) error
+	Bulk(ctx context.Context, indexName string, docs []interface{}) error
+}
+
+func NewBlockchainIndexer(ethClient EthereumClient, esClient ElasticsearchClient, indexInterval time.Duration) *BlockchainIndexer {
+	return &BlockchainIndexer{
+		ethClient:     ethClient,
+		esClient:      esClient,
+		blockChan:     make(chan *types.Block, 100),
+		txChan:        make(chan *types.Transaction, 100),
+		indexInterval: indexInterval,
+		logger:        slog.Default(),
+	}
+}
+
+func (bi *BlockchainIndexer) Start(ctx context.Context) error {
+	if err := bi.esClient.CreateIndices(ctx); err != nil {
+		return fmt.Errorf("failed to create indices: %w", err)
+	}
+
+	latestBlockNumber, err := bi.ethClient.BlockNumber(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get latest block number: %w", err)
+	}
+
+	if err = bi.initializeForwardIndex(ctx, latestBlockNumber.Uint64()); err != nil {
+		return err
+	}
+
+	if err = bi.initializeBackwardIndex(ctx, latestBlockNumber.Uint64()); err != nil {
+		return err
+	}
+
+	go bi.fetchForwardBlocks(ctx)
+	go bi.fetchBackwardBlocks(ctx)
+	go bi.processBlocks(ctx)
+
+	// Block the main function indefinitely
+	select {}
+}
+
+func (bi *BlockchainIndexer) initializeForwardIndex(ctx context.Context, latestBlockNumber uint64) error {
+	lastForwardIndexedBlock, err := bi.esClient.GetLastIndexedBlock(ctx, "forward")
+	if err != nil {
+		return fmt.Errorf("failed to get last forward indexed block: %w", err)
+	}
+	bi.logger.Info("last indexed block", "blockNumber", lastForwardIndexedBlock, "direction", "forward")
+
+	if lastForwardIndexedBlock == nil || lastForwardIndexedBlock.Cmp(big.NewInt(0)) == 0 {
```

Comment:
There is no need to check if the `lastForwardIndexedBlock` is nil, by convention it should be zero value. The whole check can be simplified to:
```Golang
if lastForwardIndexedBlock.Sign() == 0 {
// ...
```
The same goes for the `initializeBackwardIndex` function below.